src/NodeApi/JSDispatcherQueue.cs

```diff
@@ -0,0 +1,221 @@
+
+            // Under lock see if we have more tasks, complete shutdown, or start waiting.
+            lock (_queueMutex)
+            {
+                // Swap reader and writer queues.
+                (_readerQueue, _writerQueue) = (_writerQueue, _readerQueue);
+
+                if (_readerQueue.Count > 0)
+                {
+                    // We have more work to do. Start the loop from the beginning.
+                    continue;
+                }
+
+                if (_onShutdownCompleted != null && _deferralCount == 0)
+                {
+                    // Complete the shutdown: the shutdown is already started,
+                    // there are no deferrals, and all work is completed.
+                    _isShutdownCompleted = true;
+                    break;
+                }
+
+                // Wait for more work to come.
+                Monitor.Wait(_queueMutex);
+            }
+        }
+
+        // Notify about the shutdown completion.
+        ShutdownCompleted?.Invoke(this, null);
+        _onShutdownCompleted.SetResult(0);
+    }
+
+    // Create new Deferral and increment deferral count.
+    internal JSDispatcherQueueDeferral CreateDeferral()
+    {
+        lock (_queueMutex)
+        {
+            _deferralCount++;
+        }
+
+        return new JSDispatcherQueueDeferral(() =>
+        {
+            // Decrement deferral count upon deferral completion.
+            TryEnqueue(() => _deferralCount--);
+        });
+    }
+
+    internal void Shutdown(TaskCompletionSource<int> completion)
+    {
+        // Try to start the shutdown process.
+        bool isShutdownStarted = TryEnqueue(() =>
+        {
+            if (_onShutdownCompleted != null)
+            {
+                // The shutdown is already started. Subscribe to its completion.
+                ShutdownCompleted += (_, _) => completion.SetResult(0);
+                return;
+            }
+
+            // Start the shutdown process.
+            _onShutdownCompleted = completion;
+            ShutdownStarting?.Invoke(
+                this, new DispatcherQueueShutdownStartingEventArgs(() => CreateDeferral()));
+        });
+
+        if (!isShutdownStarted)
+        {
+            // The shutdown was already completed.
+            completion.SetResult(0);
+        }
+    }
+
+    private struct CurrentQueueHolder : IDisposable
+    {
+        private readonly JSDispatcherQueue? _previousCurrentQueue;
+
+        public CurrentQueueHolder(JSDispatcherQueue queue)
+        {
+            _previousCurrentQueue = s_currentQueue;
+            s_currentQueue = queue;
+            queue._threadId = Environment.CurrentManagedThreadId;
+        }
+
+        public void Dispose()
+        {
+            if (s_currentQueue != null)
+            {
+                s_currentQueue._threadId = default;
+            }
+
+            s_currentQueue = _previousCurrentQueue;
+        }
+    }
+}
+
+
+public class JSDispatcherQueueController
```

Comment:
Can these 2 methosd be moved into `JSDispatcherQueue`? I don't see why it's a separate class.

Reply:
It is a separation between a queue and its owner. The queue can be passed around, but only the owner has the right to shutdown it. Splitting up the responsibility/scope is a good thing. E.g. compare `CancellationToken` vs `CancellationTokenSource`.